### DISCUSSION TITLE
How to run deno in an airgapped environment? Can I package deno applications to be self contained and not require internet access?

### DISCUSSION THREAD

**Discussion by AudriusButkevicius** [denoland/deno#21973] [Q&A]
I work in an air-gapped environment, where I, as a user have no access to the internet, not even via a proxy.

For development, we run JFrog Artifactory which has access to the internet via a proxy, which is effectively a registry (pypi, npm, dockerhub, nuget, etc) mirror. It virus scans artefacts on ingestion.

This means I can still `pip/npm install` things off the internet to do development.

Now one of the tools that I'm trying to run internally ships it's CLI via deno, and the instructions are `deno install https://deno.land/blah.ts`.

I have no idea what that does behind the scenes or what the transport is for deno.
Hence my questions, with follow up questions

Does deno just use npm behind the scenes, and I could in theory just setup a npm proxy in Artifactory, pointing it to deno.land? 
If yes:
- How do I point deno at artifactory?
- Will it be smart enough to rewrite the dependencies too? i.e., instead of `deno run https://deno.land/blah` presumably I now do `deno run https://artifactory.local/blah` which now imports `https://deno.land/foo`, does that still work etc.

If no:
- Is Deno protocol unique enough that effectively nothing in artifactory can be leveraged for it?

Can I somehow outside of the air gapped environment package a deno application, so that it's self contained and deno does not need to reach out to the internet to run it?

Many thanks.

**Comment by Code2Life** [comment]
Both JSR & deno.land are not supported by JFrog, I have to avoid using @std or any none npm packages. It's really painful.

**Comment by jupegarnica** [comment]
you can compile an standalone executable:
https://docs.deno.com/runtime/manual/tools/compiler
